# Исследование надёжности заёмщиков по данным банка

## Импорт необходимых библиотек

In [ ]:
import pandas as pd

## Первичный анализ данных

In [ ]:
data = pd.read_csv('/datasets/data.csv')

In [ ]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


После нашего первичного знакомства с таблицей, мы моем сделать следующие выводы:

1. В столбце "total_income" и  "days_employed" колличество строк гораздо меньше, чем в остальных. Это говорит о том, что в данном столбце есть какие-то проблемы с данными.
2.  Некорректно отображается минимальный и максимальный показатель по детям(ну 20 детей вполне вероятно, что такое возможно, но не будем упускать этот момент из виду, но -1 ребёнок...)
3. Так же огромный разброс в колличестве стажа работы. Данный стоблец имеет отрцательные показатели так и сверхположительные. Максимальное значение 400 тыс. дней, что равно ~100 годам работы при максмальном занчении в столбце "dob_years" в 75 лет. Что явно говорит о том, что данные отображаются неккоректно.
4. Возраст. Минимальное значение равняется 0. Тут два варианта. Либо он не указан, либо какая-то ошибка.
5. В столбце education значения имеют разный регистр, но одинаковый id.
6. в столце purpose имеются одинаковые по смыслу цели, но разные по описанию
7. столбец days_employed имеет тип float64, что не совсем корректно для наших исследований, т.к. нельзя сказать, что человек отработал 25.62737237 дней

## Предобработка данных

### Удаление пропусков

In [ ]:
data.isna().sum() # Изучим кол-во пропусков в таблице

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Как видим в двух столбцах есть пропущенные значения. Один из них — `days_employed`. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце следует медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

Заполним пропуски в столбце days_employed медианными значениями для каждого типа занятости income_type.

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

Убедимся, что все пропуски заполнены

In [ ]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Обработка аномальных значений

Во время первичного анализа данных, нами была выявлена аномалия -  отрицательное количество дней трудового стажа в столбце `days_employed`. Это явная ошибка в данных, обработаем значения в этом столбце, заменив все отрицательные значения положительными с помощью метода `abs()`.

In [ ]:
data['days_employed'] = data['days_employed'].abs()

Проверим наличие ошибок в и столбце `children`, выведя список уникальных значений в этом столбце.

In [ ]:
data['children'].unique()  #  Выведем перечень уникальных значений столбца `children`.

array([ 1,  0,  3,  2, -1,  4, 20,  5])

В столбце `children` есть два аномальных значения. Удалим эти строки из датафрейма

In [ ]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

Убедимся, что артефакты удалены

In [ ]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Изменение типов данных

Замените вещественный тип данных в столбце `total_income` и `days_employed` на целочисленный с помощью метода `astype()`.

In [ ]:
data['total_income'] = data['total_income'].astype(int)
data['days_employed '] = data['total_income'].astype(int)

### Обработка дубликатов

Выведем на экран количество строк-дубликатов в данных. И если такие строки присутствуют, удалим их.

In [ ]:
data.duplicated().sum()

54

In [ ]:
data = data.drop_duplicates()

Обработаем также и неявные дубликаты в столбце `education` найденные на этапе первичного анализа. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведем их к нижнему регистру.

In [ ]:
data['education'] = data['education'].str.lower()

### Категоризация данных

Для облегения дальнейшего анализа, создадим в датафрейме `data` столбец `total_income_category` с категориями со следующими:

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

In [ ]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [ ]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

Выведем на экран перечень уникальных целей взятия кредита из столбца `purpose`.

In [ ]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Как мы можем, видеть, по сути все указанные выше цели, можно отнести к 4м основным категориям:

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category` с категориями выделенными мной выше.

In [ ]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [ ]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследование гипотез

Теперь подробнее изучим гипотезы, поставленных в самом начале исследования:
- Есть ли зависимость между количеством детей и возвратом кредита в срок?
- Есть ли зависимость между семейным положением и возвратом кредита в срок?
- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
- Как разные цели кредита влияют на его возврат в срок?


**Есть ли зависимость между количеством детей и возвратом кредита в срок?**

In [ ]:
(data.groupby('children')['debt'].sum() / data.groupby('children')['debt'].count())*100

children
0    7.535266
1    9.232689
2    9.454191
3    8.181818
4    9.756098
5    0.000000
Name: debt, dtype: float64

**Вывод:**

 Колличество детей не сильно влияет на вероятность своевременной выплаты кредита. Из всех результатов, наименьший процент задолженностей по кредитам (7.535266 %) показали только люди без детей, далее результаты были примерно равны, за исключением семей с тремя детьми, также полное отсутвие задолженности показали люди с пятью детьми, но я считаю что выборка таких людей слишком мала (9 человек) чтобы делать окончательные выводы.

 Таким образом, можно точно заявить что отсутсвие детей, положительно влияет на возврат кредита в срок. Далее, незаивисмо от колличества детей, участники выборки показывают одинаковый  процент просрочек по кредитам с минимальными расхождениями, кроме семей с тремя детьми (8.181818 %) где этот процент понижается.

Лучший заёмщик- без детей, Худший заёмщик- 4 ребенка

**Есть ли зависимость между семейным положением и возвратом кредита в срок?**

In [ ]:
(data.groupby('family_status')['debt'].sum() / data.groupby('family_status')['debt'].count())*100 # Ваш код будет здесь. Вы можете создавать новые ячейки.

family_status
Не женат / не замужем    9.763948
в разводе                7.064760
вдовец / вдова           6.624606
гражданский брак         9.286059
женат / замужем          7.557476
Name: debt, dtype: float64

**Вывод:** Да и разница довольно ощутима, так например не связанные узами брака люди (9.763948 % людей с просроченным долгом от общего числа людей в базе), намного чаще, чем вдовцы (6.624606 % людей с просроченным долгом от общего числа людей в базе) не выплачивают долг вовремя. В то время, как люди состоящие в браке, обладают меньшим колличеством, просроченных кредитов, несмотря на то, что являются одной из самых крупных групп в выборке. Похоже, что люди не связанные браком сейчас и никогда не бывшие в браке ('Не женат / не замужем'  и 'гражданский брак') чаще склонны не возвращать кредит в срок.

Лучший заёмщик- Вдовец, Худший заёмщик- Не женатый

**Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

In [ ]:
data.groupby('total_income_category')['debt'].count() # Ваш код будет здесь. Вы можете создавать новые ячейки.

total_income_category
A       25
B     5014
C    15938
D      349
E       22
Name: debt, dtype: int64

In [ ]:
data.groupby('total_income_category')['debt'].sum()

total_income_category
A       2
B     354
C    1353
D      21
E       2
Name: debt, dtype: int64

In [ ]:
(data.groupby('total_income_category')['debt'].sum() / data.groupby('total_income_category')['debt'].count())*100

total_income_category
A    8.000000
B    7.060231
C    8.489145
D    6.017192
E    9.090909
Name: debt, dtype: float64

**Вывод:** Согласно показаниям, прямой зависимости между уровнем дохода и вероянтностью возврата кредита в срок выявить не удалось. Так например категория D ( уровень дохода 30001–50000), показал меньший процент невозвращенных кредитов, чем категории с большим доходом


В то-же время, категория B, с большим доходом чем категория С, показала меньший уровень просрочек по кредитам, чем другие категории равно как с большим (A) так и меньшим доходом (С). Таким образом наименьший процент просроченных кредитов был у групп D (6.017192 %) и затем С (7.060231 %). Остальные группы были примерно равны, за исключением группы Е с наибольшим результатом просроченных кредитов.


Я считаю, что выборка по категориям А и Е слишком мала, чтобы делать окончательные выводы по зависимости уровня дохода от вероятности возврата кредита в срок по этим двум группам, хотя эта закономерность четко прослеживается.


Луч- D (30001–50000) Худ- E (0–30000)

**Как разные цели кредита влияют на его возврат в срок?**

In [ ]:
data.groupby('purpose_category')['debt'].count()

purpose_category
операции с автомобилем       4281
операции с недвижимостью    10754
получение образования        3989
проведение свадьбы           2324
Name: debt, dtype: int64

In [ ]:
(data.groupby('purpose_category')['debt'].sum() / data.groupby('purpose_category')['debt'].count())*100

purpose_category
операции с автомобилем      9.343611
операции с недвижимостью    7.253115
получение образования       9.250439
проведение свадьбы          7.874355
Name: debt, dtype: float64

**Вывод:** Наибольшее число целей получения кредита демонстрирует цель- операции с недвижимостью, они также являются и наибольшими (7.253115%) по вероятности возвращения кредита, второй по вероятности возврата кредита в срок, является цель- проведение свадьбы с вторым по вероятности возврата кредита (7.874355 %).
Луч- операции с недвижимостью Худ- операции с автомобилем Луч- D (30001–50000) Худ- E (0–30000) Луч- Вдовец, Худ- Не женатый Луч- без детей, Худ- 4 ребенка  

## Общий вывод.

1.  Люди у которых нет детей более ответственны к своим долговым обязательствам, чем те, у кого они есть.
2.  Люди, не имеющие законных отношений более склонны к допущению задолжности.
3. Чем ниже достаток, тем более ответственен заёмщик
4. Люди, чья цель кредита является в приобретении недвижимости являются самыми надёжными, чем те чья цель является приобретение автомобиля или получение образования



**Заключение**


Получив ответы на вопросы выше, мы можем сделать вывод, что самым надежным заёмщиком , будет клиент, отвечающий следующим критериям:
    1. Вдовец, без детей, с доходом от 30001 до 50000, берущий кредит на операции с недвижимостью

В то время как наименее надежным заёмщиком, будет:
    1. Не женатый человек, с 4мя детьми и доходом от 0 до 30000, берущий кредит для операций с автомобилем или для получения образования